rebuild/src/functions.cc

```diff
@@ -0,0 +1,775 @@
+            Napi::TypeError::New(
+                env, "BLST_ERROR: msgs and publicKeys must be the same length")
+                .ThrowAsJavaScriptException();
+            return env.Undefined();
+        }
+
+        std::unique_ptr<blst::Pairing> ctx{
+            new blst::Pairing(true, module->_dst)};
+        for (uint32_t i = 0; i < pk_array_length; i++) {
+            Napi::Value msg_value = msgs_array[i];
+            BLST_TS_UNWRAP_UINT_8_ARRAY(msg_value, msg, "msg")
+
+            Napi::Value pk_val = pk_array[i];
+            P1AffineGroup pk_point;
+            if (pk_val.IsTypedArray()) {
+                Napi::Uint8Array typed_array = pk_val.As<Napi::Uint8Array>();
+                std::string err_out{"BLST_ERROR: PublicKeyArg"};
+                if (!is_valid_length(
+                        err_out,
+                        typed_array.ByteLength(),
+                        BLST_TS_PUBLIC_KEY_LENGTH_COMPRESSED,
+                        BLST_TS_PUBLIC_KEY_LENGTH_UNCOMPRESSED)) {
+                    Napi::TypeError::New(env, err_out)
+                        .ThrowAsJavaScriptException();
+                    return env.Undefined();
+                } else if (is_zero_bytes(
+                               typed_array.Data(),
+                               0,
+                               typed_array.ByteLength())) {
+                    Napi::TypeError::New(
+                        env, "BLST_ERROR: PublicKeyArg must not be zero key")
+                        .ThrowAsJavaScriptException();
+                    return env.Undefined();
+                } else {
+                    pk_point.smart_pointer = std::make_unique<blst::P1_Affine>(
+                        typed_array.Data(), typed_array.ByteLength());
+                    pk_point.raw_point = pk_point.smart_pointer.get();
+                }
+            } else {
+                PublicKey *to_verify =
+                    PublicKey::Unwrap(pk_val.As<Napi::Object>());
+                pk_point = to_verify->_point->AsAffine();
+            }
+
+            blst::BLST_ERROR err = ctx->aggregate(
+                pk_point.raw_point,
+                sig_point.raw_point,
+                msg.Data(),
+                msg.ByteLength());
+            if (err != blst::BLST_ERROR::BLST_SUCCESS) {
+                std::ostringstream msg;
+                msg << "BLST_ERROR::" << module->GetBlstErrorString(err)
+                    << ": Invalid verification aggregate at index " << i;
+                Napi::Error::New(env, msg.str()).ThrowAsJavaScriptException();
+                return env.Undefined();
+            }
+        }
+
+        ctx->commit();
+        blst::PT pt{*sig_point.raw_point};
+        return Napi::Boolean::New(env, ctx->finalverify(&pt));
+    } catch (...) {
+        return Napi::Boolean::New(env, false);
+    }
+}
+
+typedef struct {
+    P1AffineGroup pk_point;
+    uint8_t *msg;
+    size_t msg_len;
```

Comment:
Ahhh, thats a copy of the pointer to a JSValue.  The JS runtime manages that object.  There is a nuance of async node that the Isolate (the virtual machine instance) is not available unless on the main thread.  In this context that data is being consumed on a worker thread for concurrent processing so a copy of the information stored in the isolate needs to be made so it can be referenced off-main-thread.

non-async version
```c++
blst::BLST_ERROR err = ctx->aggregate(
    pk_point.raw_point,
    sig_point.raw_point,
    msg.Data(),
    msg.ByteLength());
```

async version
```c++
Napi::Value msg_value = msgs_array[i];
BLST_TS_CLASS_UNWRAP_UINT_8_ARRAY(msg_value, msg, "msg")
m_sets[i].msg = msg.Data();
m_sets[i].msg_len = msg.ByteLength();
```

Reply:
Although I did just notice that this iteration of the refactor does not include a ref to ensure the source buffer cannot be garbage collected accidentally (although unlikely to happen).  So happy days and another great questions for a reason you didnt even realize!!! You rock dude!

Reply:
updated refs in ed48ba50a8689bbbc9e7c651e4a38f0801c39a0b